flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.executiongraph.failover.flip1;
+
+import org.apache.flink.runtime.JobException;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.throwable.ThrowableClassifier;
+import org.apache.flink.runtime.throwable.ThrowableType;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * This handler deals with task failures to return a {@link FailureHandlingResult} which contains tasks
+ * to restart to recover from failures.
+ */
+public class ExecutionFailureHandler {
+
+	/** Strategy to judge which tasks should be restarted. */
+	private final FailoverStrategy failoverStrategy;
+
+	/** Strategy to judge whether and when a restarting should be done. */
+	private final RestartBackoffTimeStrategy restartBackoffTimeStrategy;
+
+	/**
+	 * Creates the handler to deal with task failures.
+	 *
+	 * @param failoverStrategy helps to decide tasks to restart on task failures
+	 * @param restartBackoffTimeStrategy helps to decide whether to restart failed tasks and the restarting delay
+	 */
+	public ExecutionFailureHandler(
+		FailoverStrategy failoverStrategy,
+		RestartBackoffTimeStrategy restartBackoffTimeStrategy) {
+
+		this.failoverStrategy = checkNotNull(failoverStrategy);
+		this.restartBackoffTimeStrategy = checkNotNull(restartBackoffTimeStrategy);
+	}
+
+	/**
+	 * Return result of failure handling. Can be a set of task vertices to restart
+	 * and a delay of the restarting. Or that the failure is not recoverable and the reason for it.
+	 *
+	 * @param failedTask is the ID of the failed task vertex
+	 * @param cause of the task failure
+	 * @return result of the failure handling
+	 */
+	public FailureHandlingResult getFailureHandlingResult(ExecutionVertexID failedTask, Throwable cause) {
+		if (ThrowableClassifier.getThrowableType(cause) == ThrowableType.NonRecoverableError) {
```

Comment:
according to the [scheduler interface draft](https://docs.google.com/document/d/1fstkML72YBO1tGD_dmG2rwvd9bklhRVauh4FSsDDwXU/edit#), the `RestartBackoffTimeStrategy` is responsible for this decision.

```The delay is computed by a RestartBackoffTimeStrategy that is also responsible for deciding whether a restart should be performed at all (e.g., if the failure cause is SuppressRestartsException).```

Reply:
Yes the design doc proposed to check the error type in the restart strategy. But that requires every strategy implementation to do the same check.

Therefore I think it's better we let the restart strategy do it's dedicated check only. 

Besides, I'm also thinking that whether we can have a composed restart strategy, which can do multiple checks for failing max limit, failing rate, etc.